Let user define custom 404 page

Arcath opened the pull request:
I use static-server in my [Jekyll-Atom](https://github.com/Arcath/jekyll-atom) plugin to host a preview of the Jekyll site. Something I want to try needs my preview server to send 404 errors to my 404.html (like Github Pages does).

This PR adds the `error404page` option to static-server so that users can define how to handle 404 errors instead of just sending a 404 status.

yanickrochon commented:
Also, please follow the coding style already in use. You are missing some semi-colons and spaces around parenthesizes..

Arcath commented:
I have added the missing semi-colons.

It doesn't loop, it calls `getFileStats` again for the 404 page and if found sends it, if not it runs a copy of the same code as before. I've added a test to show this.

nbluis commented:
Nice @Arcath thanks for the contribution, and thank you again for your support @yanickrochon.
Once I have time I will review and publish it.
:smiley: :smiley: :smiley:

Arcath commented:
Just moved everything out into its own function so are no anonymous callbacks defined inside anonymous callbacks.

Hopefully a bit neater and easier to edit in the future if we want to handle more errors gracefully.

Arcath commented:
slightly embarrassing there, ran the tests and they passed but I'm not sure how that could have happened, fixed now.

yanickrochon commented:
Well, I have a thing with [magic numbers](http://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad). Even if the HTTP status codes are fixed, they are what they are, and are meant for programs. This is why I specified them in `const`.

Because I am in middle of other projects, this one has been put in lower priority, but I had the idea of implementing such thing, and refactor some part of the code, too.

My idea was to refactor the `options.index` value into a `options.templates` object. Such that

```
var server = new StaticServer({
  rootPath: '.',            // required, the root of the server file tree
  name: 'my-http-server',   // optional, will set "X-Powered-by" HTTP header
  port: 1337,               // optional, defaults to a random port
  host: '10.0.0.100',       // optional, defaults to any interface
  followSymlink: true,      // optional, defaults to a 404 error
  templates: {
    index: 'custom_index.html',   // optional, defaults to 'index.html'
    notFound: '404.html',         // optional, defaults to simple text message
    // etc.
});
```

Where each template may be a `string` (expects to be a file path) or a `function` receiving the `req`, `stat`, `err`, etc. arguments (when applicable) and returning a `string` that will be returned to the client (i.e. so templating engines may be used for rendering)

yanickrochon commented:
You might consider [squashing](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) your commits there, too.

Arcath commented:
I like the template idea. Want me to implement it for the notFound page and then index can be moved into it after this is pulled?

yanickrochon commented:
Sure! If you like the idea :smile:

Arcath commented:
Its now done though the templates objects

```javascript
'templates': {
  'notFound': '404.html'
}
```

and it can be set from the CLI through `-n` or `--not-found`

Everything is in for moving `index` to `templates.index` as well.

yanickrochon commented:
Yeah! Good job! You'll still have to wait for @nbluis to review, though, as I am not a member in this repo; I just own a (soon to be outdated) fork of it :smile:

nbluis commented:
Hi guys, excuse my delay but I'm on vacation these days. :tropical_drink: :tropical_drink: 
I am very happy with all your cooperation.

@Arcath The implementation was very good, I have only one caveat.

I think the template response should maintain the status code of the response (404 in this case). Currently as it uses the shared code of senfFile function, but it it will always return 200 in a request that should have code 404.

![screen shot 2016-03-14 at 8 22 17 pm](https://cloud.githubusercontent.com/assets/24320/13763106/781948be-ea22-11e5-85f6-6672c41df6fd.png)

What do you think ? Any thought @yanickrochon ?

yanickrochon commented:
That's interesting. If this is the case, then this PR does not introduce this erroneous behavior as the code is simply a wrapper around the existing code.

I find this funny. WIll try to reproduce and find a solution if no one else does first :smile:

yanickrochon commented:
Alright, here is a quick diff of a working solution. It assumes that, if `res.status` is not set, then set it, otherwise let it be set.

```
diff --git a/server.js b/server.js
index 87f6eca..d6ef9f3 100644
--- a/server.js
+++ b/server.js
@@ -173,15 +173,16 @@ Currently assumes that the only error would be a 404 error.
 @param err {Object} the error to handle
 */
 function handleError(server, req, res, err){
-  if(server.templates.notFound){
+  if (server.templates.notFound) {
     getFileStats(server, [server.templates.notFound], function(err, stat, file, index){
-      if(err){
+      if (err) {
         sendError(server, req, res, null, HTTP_STATUS_NOT_FOUND);
-      }else{
+      } else {
+        res.status = HTTP_STATUS_NOT_FOUND;
         sendFile(server, req, res, stat, file);
       }
     });
-  }else{
+  } else {
     sendError(server, req, res, null, HTTP_STATUS_NOT_FOUND);
   }
 }
@@ -485,10 +486,12 @@ function sendFile(server, req, res, stat, file) {
           }
         }).on('open', function (fd) {
           if (!headersSent) {
-            if (range) {
-              res.status = HTTP_STATUS_PARTIAL_CONTENT;
-            } else {
-              res.status = HTTP_STATUS_OK;
+            if (!res.status) {
+              if (range) {
+                res.status = HTTP_STATUS_PARTIAL_CONTENT;
+              } else {
+                res.status = HTTP_STATUS_OK;
+              }
             }
             res.writeHead(res.status, res.headers);
             headersSent = true;
```

Arcath commented:
@nbluis Good point, should be sorted in the latest commit. I implemented @yanickrochon's solution. The test is also now testing that a 404 is sent not a 200.

nbluis commented:
Nice, thank you guys.
I will pack and release very soon.

nbluis commented:
Done. :facepunch: 
Now I can return to my pool. :smile:

Arcath commented:
And now jekyll-atom uses 2.0.2 (https://github.com/Arcath/jekyll-atom/commit/a9394195dd0de5fdeb39acf6ce6bd98aad9f1878)
